[Species] InterPersonal Conflict

Pspritechologist opened the pull request:
# Description
Adds IPCs!
Everyone's favourite Positronic species finally available to be played in Space Station 14!

# TODO
- [x] Make the basis of the SiliconCharge system
- [x] Make a working IPC species
- [x] Improve upon Borg
- [x] Add chargers and charging systems
- [x] Localize all strings
- [ ] ~~Make a hud/shader effect for power drain~~
- [ ] ~~Make sure IPCs can be refilled (blood transfusions)~~ Seperate pr #172
- [x] Make sure IPCs can't become zombies/suffer from many other biological issues (will likely need testing)
- [x] Make chargers take machine parts for upgrading
- [ ] ~~Make Silicon batteries upgradable(?)~~ Gonna probably just wait on bodysystem.
- [x] Ensure death, crit, and revival work as intended
- [x] Fix skin color being weird
- [x] Make sure BatteryDrinker is clean

Ready to merge?


# Changelog

:cl:
- add: Added IPC to the game as a new species! Your favourite robot people are finally around to suffer like the rest of us. Be sure to stop by Engineering for a charge!

Pspritechologist committed:
Added a new mob typed named Hologram, set up a clone of the Drone whitelist system for "hardlight

Pspritechologist committed:
adds early scutters, and the basis of a charging system.

Pspritechologist committed:
Working on it

Pspritechologist committed:
Lots more work towards holograms, probably too long without committing :P

Pspritechologist committed:
Committing more

Pspritechologist committed:
More work on Holograms. Moved most things over to events, bug where hologram is deleted upon spawning and crashes as no hologram component is found?

Pspritechologist committed:
Fixed holocorgi ID

Pspritechologist committed:
Fuckin working again. Stuff is networked now, though not to any real benefit. There's a grace period upon leaving a projector of one 24th of a second before you get returned.

Pspritechologist committed:
Prepared for projected, lightbee, and hardlight holos by separating the protos, make a temp new tag Light, deleted unused files, renamed HoloData to HoloMind, and made HoloData struct to hold info on the Holo like type and whether it's hardlight. Needs implementing in YAML.

Pspritechologist committed:
Removed some unneeded things, cleaned some files, made it Holograms can interact with light things, and non-Holograms can't.

Pspritechologist committed:
Added damage sets for Holograms, kinda fixed rotation on spawn, begun work on hologram data in the component (no structs ;-;), and refined the variations of Holograms.

Pspritechologist committed:
Absolute overhaul. Silicon component is set up with an abundance of variables designed to allow for a wide variety of silicons in the future, such as Drones, Borgs, and IPCs. Charging system is simple right now, but with lots of foundation to improve. Currently takes into account ambient temperature for power drain with added risk of fire, and updates movement speed in line with charge state. Alert for battery level also added.

Temp mod doesn't work just yet due to borg not having a temperature component??

Pspritechologist committed:
Functioning IPC species! Still needs to be fine tuned, as it takes a lot off human at the moment. Charger and crit systems still to come.

Pspritechologist committed:
Far more work on IPCs as a race, some sprites for future use, unfucked the dumbass temperature alert thing (>:(), overheating fire still acting strange?

Pspritechologist committed:
Added silicon chargers in the form of a charging pad and industrial charger. Chargers heat up mobs, or even contained air as they push power through to them.

IPCs as a race yet further to finishing. Added antenna and screen markings for them.

Pspritechologist committed:
Finally found the positronic brain sprites. Temp usage for now.

Pspritechologist committed:
Fix posi brain sprite

Pspritechologist committed:
Fixed charger burn damage

Pspritechologist committed:
Gutted Holograms from the branch to break up the PRs.

Pspritechologist committed:
Fixed the dumbass colours on IPCs and set Dionas to use None (because hues looked awful on them anyway)

Pspritechologist committed:
Working on IPCs falling asleep when de powered. Sleep sucks

Pspritechologist committed:
Chargers are awesome now! Recursive check to find all items on an entity and charge them together, and working towards a visualizer. Have previously managed to get it to work, although it doesn't in this commit, as I'm still trying to get it to be unlit.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,107 @@
+    // [Dependency] private readonly FlammableSystem _flammableSystem = default!;
+    // [Dependency] private readonly PopupSystem _popup = default!;
+    // [Dependency] private readonly IGameTiming _gameTiming = default!;
+    // [Dependency] private readonly MovementSpeedModifierSystem _movementSpeedModifierSystem = default!;
```
//

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,107 @@
+
+    private void OnSiliconChargeStateUpdate(EntityUid uid, SiliconDownOnDeadComponent siliconDeadComp, SiliconChargeStateUpdateEvent args)
+    {
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,107 @@
+
+    private void SiliconDead(EntityUid uid, SiliconDownOnDeadComponent siliconDeadComp, BatteryComponent batteryComp)
+    {
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,107 @@
+            return;
+
+        var sleepComp = EntityManager.EnsureComponent<SleepingComponent>(uid);
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,107 @@
+        var sleepComp = EntityManager.EnsureComponent<SleepingComponent>(uid);
+
+        RemComp<SpamEmitSoundComponent>(uid); // This is also fucking stupid, I once again hate the sleeping system.
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/Bed/BedSystem.cs:
```diff
@@ -79,7 +79,7 @@ public override void Update(float frameTime)
                     var damage = bedComponent.Damage;
 
-                    if (HasComp<SleepingComponent>(healedEntity))
+                    if (EntityManager.TryGetComponent<SleepingComponent>(healedEntity, out var sleepComp) && sleepComp.Heal)
```
‫

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly MovementSpeedModifierSystem _movementSpeedModifierSystem = default!;
+
+    static string popupOverheating = Loc.GetString("silicon-system-overheating");
```
unnecessary static definition

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+        base.Initialize();
+
+        // Subscribe for init on entities with both SiliconComponent and BatteryComponent
+        SubscribeLocalEvent<BatteryComponent, ComponentInit>(OnBatteryInit);
```
```suggestion
        // I hate dual subscriptions
        SubscribeLocalEvent<BatteryComponent, ComponentInit>(OnBatteryInit);
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+            // If the silicon is not battery powered, or is dead, skip it.
+            if (!siliconComp.BatteryPowered ||
+                _mobStateSystem.IsDead(silicon))
+                continue;
```
could use a .Where() on the query

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+
+            var drainRate = 10 * (siliconComp.DrainRateMulti);
+
+            // All multipliers will be - 1, and then added together, and then multiplied by 150. This is then added to the base drain rate.
```
```suggestion
            // All multipliers will be subtracted by 1, and then added together, and then multiplied by the drain rate. This is then added to the base drain rate.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+            // This is to stop exponential increases, while still allowing for less-than-one multipliers.
+            var drainRateFinalAddi = 0f;
+
+            //TODO: Devise a method of adding multis where other systems can alter the drain rate.
```
```suggestion
            // TODO: Devise a method of adding multis where other systems can alter the drain rate.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+        if (!EntityManager.TryGetComponent<TemperatureComponent>(silicon, out var temperComp) ||
+            !EntityManager.TryGetComponent<ThermalRegulatorComponent>(silicon, out var thermalComp))
+        {
+            // DebugTools.Assert("Silicon has no temperature component, but is battery powered.");
```
//

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+            // DebugTools.Assert("Silicon has no temperature component, but is battery powered.");
+            return 0;
+        }
+        var siliconComp = EntityManager.GetComponent<SiliconComponent>(silicon);
```
```suggestion

        var siliconComp = EntityManager.GetComponent<SiliconComponent>(silicon);
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+        // Check if the silicon is in a hot environment.
+        if (temperComp.CurrentTemperature > upperThreshHalf)
+        {
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+            var hotTempMulti = Math.Min(temperComp.CurrentTemperature / upperThreshHalf, 4);
+
+            // Logger.DebugS("silicon", $"Silicon {silicon} is overheating, multiplier is {hotTempMulti}.");
+
```
```suggestion

```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,190 @@
+                }
+
+                // Logger for the random chances.
+                // Logger.WarningS("silicon", $"Silicon {silicon} has a {Math.Clamp(temperComp.CurrentTemperature / (upperThresh * 15), 0.001f, 0.9f)} chance to catch fire, and a {Math.Clamp(temperComp.CurrentTemperature / (upperThresh * 4), 0.001f, 0.75f)} chance to popup overheating.");
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,305 @@
+                UpdateState(chargerComp.Owner, chargerComp);
+            }
+
+            var chargeRate = frameTime * chargerComp.ChargeMulti / chargerComp.PresentEntities.Count ;
```
floatin
```suggestion
            var chargeRate = frameTime * chargerComp.ChargeMulti / chargerComp.PresentEntities.Count;
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,305 @@
+        // If the given entity has a battery, charge it.
+        if (EntityManager.TryGetComponent(entity, out BatteryComponent? batteryComp) &&
+        !EntityManager.TryGetComponent<UnremoveableComponent>(entity, out var _) &&
+        batteryComp.CurrentCharge < batteryComp.MaxCharge)
```
```suggestion
            !EntityManager.TryGetComponent<UnremoveableComponent>(entity, out var _) &&
            batteryComp.CurrentCharge < batteryComp.MaxCharge)
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,305 @@
+
+    private void ChargeBattery(EntityUid entity, BatteryComponent batteryComp, float chargeRate, SiliconChargerComponent chargerComp)
+    {
+        // Do some math so a charger never charges a battery in less than 10 seconds, just for the effect of it.
```
maybe a minchargetime variable or something

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,305 @@
+        var damage = new DamageSpecifier(_prototypeManager.Index<DamageTypePrototype>("Shock"), frameTime * chargerComp.ChargeMulti / 100);
+        var damageDealt = _damageableSystem.TryChangeDamage(entity, damage, false, true, damageComp, chargerComp.Owner);
+        chargerComp.warningAccumulator -= frameTime;
+        if (damageDealt != null && chargerComp.warningAccumulator <= 0 && damageDealt.Total > 0)
```
```suggestion

        if (damageDealt != null && chargerComp.warningAccumulator <= 0 && damageDealt.Total > 0)
```

DEATHB4DEFEAT commented on Content.Shared/Bed/Sleep/SleepingComponent.cs:
//

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Components/SiliconComponent.cs:
```diff
@@ -0,0 +1,90 @@
+namespace Content.Shared.SimpleStation14.Silicon.Components;
+
+/// <summary>
+///     Component for defnining a mob as a robot.
```
```suggestion
///     Component for defining a mob as a robot.
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Systems/SharedSiliconSystem.cs:
```diff
@@ -0,0 +1,113 @@
+using Content.Shared.Alert;
+using Robust.Shared.Serialization;
+using Content.Shared.Movement.Systems;
+// using
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Systems/SharedSiliconSystem.cs:
```diff
@@ -0,0 +1,113 @@
+using Content.Shared.Movement.Systems;
+// using
+
+namespace Content.Shared.SimpleStation14.Silicon.Systems;
```
```suggestion
namespace Content.Shared.SimpleStation14.Silicon.Systems;

```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Systems/SharedSiliconSystem.cs:
```diff
@@ -0,0 +1,113 @@
+
+    private void OnRefreshMovespeed(EntityUid uid, SiliconComponent component, RefreshMovementSpeedModifiersEvent args)
+    {
+        if (component.BatteryPowered)
```
invert if

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Systems/SharedSiliconSystem.cs:
```diff
@@ -0,0 +1,113 @@
+
+            foreach (var state in speedModThresholds)
+            {
+                if (component.ChargeState >= state.Key && ((float)state.Key) > closest)
+                    closest = ((float)state.Key);
```
```suggestion
                if (component.ChargeState >= state.Key && ((float) state.Key) > closest)
                    closest = ((float) state.Key);
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Systems/SharedSiliconSystem.cs:
```diff
@@ -0,0 +1,113 @@
+    Randomized
+}
+
+// [Serializable, NetSerializable]
+// public struct ChargeStateThresholdsData
+// {
+//     public float Mid;
+//     public float Low;
+//     public float Critical;
+// }
```
//

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Body/Organs/ipc.yml:
```diff
@@ -0,0 +1,90 @@
+  components:
+  - type: Sprite
+    state: brain
+  - type: Organ
```
added by parent

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Body/Parts/ipc.yml:
```diff
@@ -0,0 +1,188 @@
+# TODO: Add descriptions (many)
+# TODO BODY: Part damage
```
//

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Player/silicon_base.yml:
```diff
@@ -0,0 +1,378 @@
+- type: entity
+  save: false
+  abstract: true
+  id: PlayerSiliconBase #for player controlled silicons
```
```suggestion
  id: PlayerSiliconBase # For player controlled silicons
```

DEATHB4DEFEAT reviewed (changes requested):
crin**g**nt

Pspritechologist committed:
Apply suggestions from code review

Co-authored-by: DEATHB4DEFEAT <77995199+DEATHB4DEFEAT@users.noreply.github.com>

Pspritechologist committed:
Cleaned up sleeping files.

Pspritechologist committed:
Cleaned up a bit, more suggestions

Pspritechologist committed:
False is default, and some suggestions by Death in SiliconChargerSystem.cs

Pspritechologist committed:
Added some regions :0

Pspritechologist committed:
component.Owner no more! >:D

Plus some more region stuff.

Pspritechologist committed:
component.Owner begone! :D (mostly)

Pspritechologist committed:
YA BOI DID IT
CHARGER VISUALS WORK NOW AND THEY WORK GREAT.

Also replaced elseif tree with a switch, and removed component.Owner.

Pspritechologist committed:
Workin' on that sweet sweet APC succ

Pspritechologist committed:
Cleaned up a bit post-rebase, fixed TintedHues validator, fixed drinker DoAfter

Pspritechologist committed:
Various small fixes, electrocution charge

Moved some files, renamed some Locs. Electrocuting a battery will now charge that battery an amount based on the damage, including IPCs (in theory).
Electrocution event now relays damage dealt, as it obviously should've.

Pspritechologist committed:
Shocked batteries now charge based on damage dealt and DamagePerWatt

Pspritechologist committed:
Merge branch 'master' of https://github.com/Park-Station/Parkstation into IPCs

Pspritechologist committed:
General minor clean up.
Fixes prediction issue with charger lights, and I think the IPC prototype is officially ready for action.
CyborgComponent is dumb... So very dumb...

Pspritechologist commented on Content.Shared/SimpleStation14/Silicon/Systems/SharedSiliconChargerSystem.cs:
```diff
@@ -0,0 +1,76 @@
+    /// <summary>
+    ///     Updates the visual and auditory state of the charger based on if it's active, and/or open.
+    /// </summary>
+    public void UpdateState(EntityUid uid, SiliconChargerComponent? component = null)
```
@DEATHB4DEFEAT Let me know if you know of anyway to use this the way I need to, without making it Public.

Pspritechologist committed:
Review suggestions

Pspritechologist committed:
Merge branch 'master' of https://github.com/Park-Station/Parkstation into IPCs

Pspritechologist committed:
General, minor cleanup.

Also commented out some IPC antenna that need separate layer colours.

Pspritechologist committed:
Merge remote-tracking branch 'remotes/origin/master' into IPCs

Likely need to fix whatever bugs are left behind, and also update the moved silicon.yml

Pspritechologist committed:
Merge branch 'master' of https://github.com/Park-Station/Parkstation into IPCs

Pspritechologist committed:
Fixed the errors and cleaned up a bit

Pspritechologist committed:
Several changes and fixes.

Cleaned up several parts of the code, fixed some merge issues (leg speed and slots), added FTLs for loc strings, improved aspects of charging, made chargers constructable using part upgrades.

Pspritechologist committed:
New IPC sprites v2

Slightly modified the head shape to fit better with pre-existing sprites, and look a little more 'blocky.' v1 was never committed.

Pspritechologist committed:
IPC sprites v3

Final version of sprites maybe, significantly lightened darker areas to allow for better colouration of parts.
Removed background legs and feet (what the fuck, limbs just don't work????)

Pspritechologist committed:
IPCs now use integrated radios!

IPCs now have blank RadioReceiver, Transmitter, and Active components which get updated by their EncryptionKeyHolderComponent (had to write a system for that).
Their encryption keys get filled in based on their job's headset with some modifications to StationSpawningSystem that I'm not super proud of...

caucasianhamburger commented:
i will watch your career with great interest

Pspritechologist committed:
IPC encryption keys now work better, and effect the SetOutfit command.

Pspritechologist committed:
Things are localized!

DEATHB4DEFEAT commented on Content.Server/Electrocution/ElectrocutionSystem.cs:
spess

Pspritechologist replied:
Huh?

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Components/BatteryDrinkerComponent.cs:
```diff
@@ -0,0 +1,41 @@
+
+    /// <summary>
+    ///     How long it takes to drink from a battery, in seconds.
+    ///     Is mutliplied by the source.
```
```suggestion
    ///     Is multiplied by the source.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Components/BatteryDrinkerComponent.cs:
```diff
@@ -0,0 +1,41 @@
+    ///     Doesn't _need_ to be localised, but come on, man.
+    /// </summary>
+    [DataField("drinkText")]
+    public string DrinkText = "aaaaaaaaaa";
```
aaaaaaaaaa

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Systems/BatteryDrinkerSystem.cs:
```diff
@@ -0,0 +1,150 @@
+
+    private bool TestDrinkableBattery(EntityUid target, BatteryDrinkerComponent drinkerComp)
+    {
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Systems/BatteryDrinkerSystem.cs:
```diff
@@ -0,0 +1,150 @@
+        if (sourceComp != null && sourceComp.MaxAmount > 0)
+        {
+            amountToDrink = MathF.Min(amountToDrink, (float) sourceComp.MaxAmount);
+        }
```
```suggestion
        if (sourceComp != null && sourceComp.MaxAmount > 0)
            amountToDrink = MathF.Min(amountToDrink, (float) sourceComp.MaxAmount);
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Systems/BatteryElectrocuteChargeSystem.cs:
```diff
@@ -0,0 +1,39 @@
+
+        var message = Loc.GetString("battery-electrocute-charge");
+
+        _popup.PopupEntity(message, uid, uid);
```
```suggestion
        _popup.PopupEntity(Loc.GetString("battery-electrocute-charge"), uid, uid);
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,113 @@
+        else if (args.ChargeState != ChargeState.Dead && siliconDeadComp.Dead)
+        {
+            SiliconUnDead(uid, siliconDeadComp, batteryComp);
+        }
```
```suggestion
        if (args.ChargeState == ChargeState.Dead && !siliconDeadComp.Dead)
            SiliconDead(uid, siliconDeadComp, batteryComp);
        else if (args.ChargeState != ChargeState.Dead && siliconDeadComp.Dead)
            SiliconUnDead(uid, siliconDeadComp, batteryComp);
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,113 @@
+        RaiseLocalEvent(uid, new SiliconChargeDeathEvent(uid, batteryComp));
+    }
+
+    private void SiliconUnDead(EntityUid uid, SiliconDownOnDeadComponent siliconDeadComp, BatteryComponent batteryComp)
```
funny name :+1:

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
```diff
@@ -0,0 +1,113 @@
+}
+
+/// <summary>
+///     A canellable event raised when a Silicon is about to go down due to charge.
```
```suggestion
///     A cancellable event raised when a Silicon is about to go down due to charge.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeDeathSystem.cs:
event in event file D:

Pspritechologist replied:
cope

Pspritechologist replied:
Nahh

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,316 @@
+
+        #region Entity Storage Chargers
+        // Check for any chargers with the EntityStorageComponent.
+        var entstorQuery = EntityQueryEnumerator<SiliconChargerComponent, EntityStorageComponent>();
```
```suggestion
        var entityStorageQuery = EntityQueryEnumerator<SiliconChargerComponent, EntityStorageComponent>();
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,316 @@
+
+
+                HandleChargingEntity(entity, chargeRate, chargerComp, uid, frameTime);
+
```
```suggestion

                HandleChargingEntity(entity, chargeRate, chargerComp, uid, frameTime);


```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,316 @@
+
+    private void ChargeBattery(EntityUid entity, BatteryComponent batteryComp, float chargeRate, SiliconChargerComponent chargerComp, EntityUid chargerUid)
+    {
+        // Do some math so a charger never charges a battery from zero to full in less than 10 seconds, just for the effect of it.
```
berryable

DEATHB4DEFEAT commented on Content.Server/Temperature/Systems/TemperatureSystem.cs:
wtf go to wiz

Pspritechologist replied:
Likely not a pr worth making at this point. Quite a few people have expressed interest in IPC, so it might make its way there anyway.

DEATHB4DEFEAT commented on Content.Shared/Humanoid/SkinColor.cs:
ffffikcs

DEATHB4DEFEAT commented on Content.Shared/Radio/EntitySystems/EncryptionKeySystem.cs:
uptream?

Pspritechologist replied:
Ditto, honestly.

DEATHB4DEFEAT commented on Resources/Locale/en-US/SimpleStation14/Prototypes/Entities/Mobs/Customization/ipcScreens.ftl:
```diff
@@ -0,0 +1,39 @@
+marking-ScreenGoggles = Goggles
+marking-ScreenExclaim = Exclaim
+marking-ScreenHeart = Heart
+marking-ScreenMonoeye = Cylops
```
```suggestion
marking-ScreenMonoeye = Cyclops
```

Pspritechologist committed:
Cleaned up tons of stuff.

Battery draining seems to be broken right now, reporting 0 charge at all times. Will fix soon.

Pspritechologist committed:
Fixed issue with IPC charge, and review suggestions.

Pspritechologist committed:
Lots of general improvements to the systems, lots of cleanup.

Clean, Medi, and Honk bots all use battery cell slots, which SiliconSystem now supports.
To support external batteries entirely sometime.

Pspritechologist committed:
Several improvements.

IPCs and Borgs now beep when drained of battery, battery drinker works with Loc, legs don't clip out of pants, skin color is better than before, IPC now die gracefully (although annoyingly similar to humans)

Pspritechologist committed:
Merge branch 'master' into IPC

Pspritechologist commented on Content.Server/Station/Systems/StationSpawningSystem.cs:
```diff
@@ -190,6 +193,40 @@ public void EquipStartingGear(EntityUid entity, StartingGearPrototype startingGe
+        // EncryptionKeyHolderComponent on it, move the keys over, and delete the headset.
+        if (TryComp<EncryptionKeyHolderComponent>(entity, out var keyHolderComp))
+        {
+            // Resolve SharedContainerSystem
```
```suggestion
```

Pspritechologist committed:
Update Content.Server/Station/Systems/StationSpawningSystem.cs

Pspritechologist commented on Content.Shared/Humanoid/SkinColor.cs:
```diff
@@ -117,10 +117,13 @@ public static bool VerifyHumanSkinTone(Color color)
     /// <returns>Tinted hue color</returns>
     public static Color TintedHues(Color color)
     {
```
```suggestion
    {
        // Parkstation-IPC-Start
```

Pspritechologist commented on Content.Shared/Humanoid/SkinColor.cs:
```diff
@@ -117,10 +117,13 @@ public static bool VerifyHumanSkinTone(Color color)
+        hsv.Z = Math.Max(hsv.Z, 0.25f);
 
-        return Color.FromHsv(newColor);
+        return Color.FromHsv(hsv);
```
```suggestion
        return Color.FromHsv(hsv);
        // Parkstation-IPC-End
```

Pspritechologist commented on Content.Shared/Humanoid/SkinColor.cs:
```diff
@@ -130,8 +133,9 @@ public static Color TintedHues(Color color)
     /// <returns>True if valid, false otherwise</returns>
     public static bool VerifyTintedHues(Color color)
     {
```
```suggestion
    {
        // Parkstation-IPC-Start
```

Pspritechologist commented on Content.Shared/Humanoid/SkinColor.cs:
```diff
@@ -130,8 +133,9 @@ public static Color TintedHues(Color color)
-        return Color.ToHsv(color).Y != .1f;
+        var hsv = Color.ToHsv(color);
+
+        return hsv.Y <= 0.6 && hsv.Z >= 0.25;
```
```suggestion
        return hsv.Y <= 0.6 && hsv.Z >= 0.25;
        // Parkstation-IPC-End
```

Pspritechologist committed:
Add Parkstation tags to skin colours file.

Pspritechologist committed:
Merge branch 'master' of https://github.com/Park-Station/Parkstation into IPCs

Pspritechologist committed:
Removed unintended changes.

Pspritechologist committed:
Adjust tv sprite

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Components/RandomBatteryChargeComponent.cs:
```diff
@@ -0,0 +1,23 @@
+    public Vector2 BatteryChargeMinMax = (1f, 1f);
+
+    /// <summary>
+    ///     True if the current charge is based on the preexisting current charge, or false if it's based on the max charge.
```
is this ture?

Pspritechologist replied:
Better?

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Power/Systems/SiliconEmitSoundOnDrainedSystem.cs:
```diff
@@ -0,0 +1,40 @@
+    private void OnAlive(EntityUid uid, SiliconEmitSoundOnDrainedComponent component, SiliconChargeAliveEvent args)
+    {
+        RemComp<SpamEmitSoundComponent>(uid); // This component is bad and I don't feel like making a janky work around because of it.
+        // If you give something the EmitSoundOnCritComponent, know that it can't have the SpamEmitSoundComponent, and any other systems that play with it will just be broken.
```
```suggestion
        // If you give something the SiliconEmitSoundOnDrainedComponent, know that it can't have the SpamEmitSoundComponent, and any other systems that play with it will just be broken.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Components/BatteryDrinkerSourceComponent.cs:
```diff
@@ -0,0 +1,27 @@
+    public float DrinkSpeedMulti = 1f;
+
+    /// <summary>
+    ///     The sound to play when the APC gets drunk from.
```
```suggestion
    ///     The sound to play when the battery gets drunk from.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,213 @@
+                continue;
+
+            // Check if the Silicon is an NPC, and if so, follow the delay as specified in the CVAR.
+            if (siliconComp.EntityType.Equals(SiliconType.Npc))
```
```suggestion
            if (siliconComp.EntityType == SiliconType.Npc)
```

Pspritechologist replied:
Can't, since EntityType is of type "Enum" and not of type SiliconType...

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargeSystem.cs:
```diff
@@ -0,0 +1,213 @@
+            // TODO: Devise a method of adding multis where other systems can alter the drain rate.
+            // Maybe use something similar to refreshmovespeedmodifiers, where it's stored in the component.
+            // Maybe it doesn't matter, and stuff should just use static drain?
+            if (!siliconComp.EntityType.Equals(SiliconType.Npc)) // Don't bother checking heat if it's an NPC. It's a waste of time, and it'd be delayed due to the update time.
```
```suggestion
            if (!siliconComp.EntityType == SiliconType.Npc) // Don't bother checking heat if it's an NPC. It's a waste of time, and it'd be delayed due to the update time.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Silicon/Charge/Systems/SiliconChargerSystem.cs:
```diff
@@ -0,0 +1,320 @@
+        else if (!HasComp<UnremoveableComponent>(entity) && // Should probably be charged by the entity holding it. Might be too small to be safe.
+            TryComp<BatteryComponent>(entity, out var batteryComp))
+        {
+            if (batteryComp.CurrentCharge < batteryComp.MaxCharge)
```
do the batteries still explode?

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Silicon/Components/SiliconChargerComponent.cs:
```diff
@@ -0,0 +1,124 @@
+    public TimeSpan WarningTime = TimeSpan.Zero;
+
+    /// <summary>
+    ///     The current parts mutliplier.
```
```suggestion
    ///     The current parts multiplier.
```

DEATHB4DEFEAT commented on Resources/Locale/en-US/revenant/revenant.ftl:
```diff
@@ -18,7 +18,7 @@ revenant-soul-finish-harvest = {CAPITALIZE(THE($target))} slumps onto the ground
 revenant-psionic-power = a spirit power
 
-#UI
+# UI
```
```suggestion
#UI
```

DEATHB4DEFEAT reviewed (changes requested):
- Need random names
- Age max change
- Remove marking points (there are no markings enabled)
- alerts-charge-desc
- Robot typing indicator
- Broken charging pad
- No guidebook entry
- Needs eye protection
- Remove ability to be psionic

Pspritechologist committed:
Merge branch 'upstream-sync-2' of https://github.com/Park-Station/Parkstation into IPCs

Pspritechologist committed:
Minor fixes post-merge

Pspritechologist committed:
Merge branch 'master' of https://github.com/Simple-Station/Parkstation into IPCs

Pspritechologist committed:
Review fixes

Pspritechologist committed:
Batteries should now blow up again

Pspritechologist committed:
Addressed remaining review issues.

Pspritechologist committed:
Missing FTL fix?

Pspritechologist committed:
Removed unneeded sleep change, added timer for coming back to life

Pspritechologist committed:
IPCs have names now

Pspritechologist committed:
Removed vscode file

DEATHB4DEFEAT reviewed (approved):
:godmode:
